### PR TITLE
Improve asynchronous host to device copy

### PR DESCRIPTION
**Pull request by okuta**
This PR implement asynchronous copy to `cupy.array`.
Please see also chainer/chainer#2238.

### CODE REVIEW

**Comment by gwtnb** [comment]
A test case failed.

```
ERROR: test_cupy_array2 (test_cuda.TestToGPU_param_0)  parameter: {'c_contiguous': True}
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/work/chainer/tests/chainer_tests/test_cuda.py", line 230, in test_cupy_array2
    y = cuda.to_gpu(x, device=1)
  File "/work/chainer/chainer/cuda.py", line 284, in to_gpu
    return cupy.array(array, copy=True)
  File "/home/user/.local/lib/python2.7/site-packages/cupy-1.0.0.1-py2.7-linux-x86_64.egg/cupy/creation/from_data.py", line 26, in array
    return core.array(obj, dtype, copy, ndmin)
  File "cupy/core/core.pyx", line 1884, in cupy.core.core.array (cupy/core/core.cpp:56572)
  File "cupy/core/core.pyx", line 1895, in cupy.core.core.array (cupy/core/core.cpp:56028)
  File "cupy/core/core.pyx", line 301, in cupy.core.core.ndarray.astype (cupy/core/core.cpp:8318)
TypeError: order not understood
```

File: cupy/core/core.pyx [hunk 1]
```diff
@@ -1885,29 +1886,34 @@ cpdef ndarray array(obj, dtype=None, bint copy=True, Py_ssize_t ndmin=0):
     cdef Py_ssize_t nvidem
     cdef ndarray a
     if isinstance(obj, ndarray):
+        a = obj
         if dtype is None:
-            dtype = obj.dtype
-        a = obj.astype(dtype, copy=copy)
+            dtype = a.dtype
+        if a.data.device.id == device.get_device_id():
+            a = a.astype(dtype, copy)
```
Comment (gwtnb) [inline cupy/core/core.pyx]: `copy=copy`

**Comment by gwtnb** [comment]
Could you resolve conflicts?

**Comment by okuta** [comment]
@gwtnb I resolved conflicts.

File: cupy/cuda/pinned_memory.pyx [hunk 2]
```diff
@@ -129,12 +130,63 @@ cdef class PinnedMemoryPointer:
...
+            self._check_and_release_without_lock()
+            if event.done:
+                return
+            self.events.append((event, obj))
+
+    cpdef check_and_release(self):
+        """ Check and release completed events.
+
+        """
+        if len(self.events) == 0:
```
Comment (gwtnb) [inline cupy/cuda/pinned_memory.pyx]: Please conform to PEP8. https://github.com/cupy/cupy/pull/123/files

> For sequences, (strings, lists, tuples), use the fact that empty sequences are false.
>
> Yes: if not seq:
>     if seq:
>
> No: if len(seq):
>    if not len(seq):

File: cupy/cuda/pinned_memory.pyx [hunk 3]
```diff
@@ -129,12 +130,63 @@ cdef class PinnedMemoryPointer:
...
+            return
+        with self.lock:
+            self._check_and_release_without_lock()
+
+    cpdef _check_and_release_without_lock(self):
+        while len(self.events) != 0 and self.events[0][0].done:
```
Comment (gwtnb) [inline cupy/cuda/pinned_memory.pyx]: ditto

**Comment by okuta** [comment]
@gwtnb I fixed.

**Comment by gwtnb** [comment]
LGTM

**Pull request merged**